src/lib/nullifier.ts

```diff
@@ -0,0 +1,106 @@
+import {
+  Poseidon,
+  PublicKey,
+  Field,
+  Group,
+  Struct,
+  MerkleMapWitness,
+  Scalar,
+} from 'snarkyjs';
+import type { Nullifier as JsonNullifier } from '../mina-signer/src/TSTypes.js';
+import { scaleShifted } from './signature.js';
+
+export { Nullifier };
+
+/**
+ * RFC: https://github.com/o1-labs/snarkyjs/issues/756
+ *
+ * Paper: https://eprint.iacr.org/2022/1255.pdf
+ */
+class Nullifier extends Struct({
+  publicKey: PublicKey,
+  public: {
+    nullifier: Group,
+    s: Scalar,
+  },
+  private: {
+    c: Field,
+    g_r: Group,
+    h_m_pk_r: Group,
+  },
+}) {
+  static fromJSON(json: JsonNullifier): Nullifier {
+    return super.fromJSON(json as any) as Nullifier;
+  }
+
+  /**
+   * Verifies that the Nullifier belongs to a specific message. Throws an error if the Nullifier is incorrect.
+   */
+  verify(message: Field[]) {
+    let {
+      publicKey,
+      public: { nullifier, s },
+      private: { c },
+    } = this;
+
+    // generator
+    let G = Group.generator;
+
+    // serialize public key into fields once
+    let pk_fields = publicKey.toFields();
+
+    // x and y of hash(msg, pk), it doesn't return a Group because y is split into x0 and x1, both two roots of a field element
+    let {
+      x,
+      y: { x0 },
+    } = Poseidon.hashToGroup([...message, ...pk_fields]);
```

Comment:
your scheme seems to be different from the original scheme in one minor aspect: you hash the public key in compressed `(x, isOdd)` form (`publicKey.toFields()`) instead of in the uncompressed curve point form `(x, y)` (`publicKey.toGroup().toFields()`). I'm sure this has no security impact, but I think it would be cleaner to use the uncompressed curve point form, because that seems to be closer to an ideal "standard implementation". Usage of the public key in compressed form is really more a Mina implementation detail that should be kept out of here.

Decompressing with `publicKey.toGroup()` is an expensive operation in the proof, so it might make sense to make this operation _optional_ for the prover by returning the _uncompressed_ publicKey (the `Group`) from the wallet. Then the public compressed pk plays no role in the entire protocol.

For smart contracts it would then be convenient to have a method `getPublicKey(): PublicKey` on `Nullifier`, so it becomes

```ts
Nullifier extends Struct({ publicKey: Group, ... }) {
   getPublicKey() {
     return PublicKey.fromGroup(this.publicKey);
   }
   // ...
}
```

Reply:
well - decide for yourself if you think this is worth it

Reply:
I would rather keep it close to the original protocol, I'll change it to use the uncompressed point instead!